sdk/core/azure-core/inc/azure/core/internal/http/pipeline.hpp

```diff
@@ -49,12 +51,123 @@ namespace Azure { namespace Core { namespace Internal { namespace Http {
       }
 
       m_policies.reserve(policies.size());
-      for (auto&& policy : policies)
+      for (auto& policy : policies)
       {
         m_policies.emplace_back(policy->Clone());
       }
     }
 
+    /**
+     * @brief Construct a new Http Pipeline object from clientOptions.
+     *
+     * @remark The client options includes per retry and per call policies which are merged with the
+     * service-specific per retry policies.
+     *
+     * @param clientOptions The SDK client options.
+     * @param telemetryServiceName The name of the service for sending telemetry.
+     * @param telemetryServiceVersion The version of the service for sending telemetry.
+     * @param perRetryPolicies The service specific per retry policies.
+     * @param authBeforeSendPolicy Optional policy to set authentication information to the request
+     * right before sending it.
+     */
+    explicit HttpPipeline(
+        ClientOptions const& clientOptions,
+        std::string const& telemetryServiceName,
+        std::string const& telemetryServiceVersion,
+        std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>>&& perRetryPolicies,
+        std::unique_ptr<Azure::Core::Http::HttpPolicy> authBeforeSendPolicy = nullptr)
+        : HttpPipeline(
+            clientOptions,
+            telemetryServiceName,
+            telemetryServiceVersion,
+            std::move(perRetryPolicies),
+            {},
+            std::move(authBeforeSendPolicy))
+    {
+    }
+
+    /**
+     * @brief Construct a new Http Pipeline object from clientOptions.
+     *
+     * @remark The client options includes per retry and per call policies which are merged with the
+     * service-specific per retry policies.
+     *
+     * @param clientOptions The SDK client options.
+     * @param telemetryServiceName The name of the service for sending telemetry.
+     * @param telemetryServiceVersion The version of the service for sending telemetry.
+     * @param perRetryPolicies The service-specific per retry policies.
+     * @param perCallPolicies The service-specific per call policies.
+     * @param authBeforeSendPolicy Optional policy to set authentication information to the request
+     * right before sending it.
+     */
+    explicit HttpPipeline(
+        ClientOptions const& clientOptions,
+        std::string const& telemetryServiceName,
+        std::string const& telemetryServiceVersion,
+        std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>>&& perRetryPolicies,
+        std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>>&& perCallPolicies,
+        std::unique_ptr<Azure::Core::Http::HttpPolicy> authBeforeSendPolicy = nullptr)
```

Comment:
This last option parameter is unnessecary.  Storage can pass in the perRetryPolicies and put their authBeforeSendPolicy as the last in that list.  That is what other languages do, why isn't that acceptable here?

Reply:
@Jinming-Hu , as per Azure Core team conversation (and also Ted and Pavel from .Net), we will not support the feature to make sure the addition of a policy to be right before the transport policy.

It looks like in all the history of storage .Net, they have never faced the issue of an end-user adding a per-retry policy which updates the request combined with the `sharedKeyAuthPolicy`.  They mention that end-users typically uses per-call policies only.